### PR TITLE
Use thiserror instead of quick-error

### PR DESCRIPTION
**Pull request by koushiro**
Signed-off-by: koushiro <koushiro.cqx@gmail.com>

- Use failure instead of quick-error
- Fix some typos
- Reorder some import

### CODE REVIEW

**Comment by kennytm** [comment]
PTAL @breeswish you suggested `failure` in https://github.com/pingcap/rust-prometheus/issues/218#issuecomment-458882638. 

However, `failure` is pretty much a failure now (its author has released `fehler` as the latest experiment), I question whether this is still a correct replacement. 

(See also https://www.reddit.com/r/rust/comments/dfs1zk/2019_q4_error_patterns_snafu_vs_errderive_anyhow/ for the latest discussion.)

**Comment by breeswish** [comment]
I agree with @kennytm that failure is not the best choice now. We have many new choices like [fehler](https://github.com/withoutboats/), [thiserror](https://github.com/dtolnay/thiserror) and [snafu](https://github.com/shepmaster/snafu). I guess thiserror should be used instead as an error used in library. What do you think? @kennytm

**Comment by koushiro** [comment]
I tried the [thiserror](https://github.com/dtolnay/thiserror) crate, which is awesome. I will use it instead of `failure` if @kennytm agrees with that.

**Comment by kennytm** [comment]
SGTM. Comparing the three, `thiserror` is the lightest, and also the most stable (in terms of semver).

`fehler` is more like an experiment (to fit exceptions into Rust). It isn't a good idea for us to depend on an unstable design pattern. 

`snafu` has quite a lot of features, and thus 3× the crate size of the other two. Like `failure` it introduced a `.context()` method and backtrace. These would be useful for TiKV, though here in `rust-prometheus` we didn't use these.

**Comment by koushiro** [comment]
@breeswish @kennytm PTAL :)

File: src/errors.rs [hunk 1]
```diff
@@ -11,76 +11,28 @@
...
-    }
+/// The error types for prometheus.
+#[derive(Debug, Error)]
+pub enum Error {
+    /// A duplicate metric collector has already been registered.
+    #[error("Duplicate metrics collector registration attempted")]
+    AlreadyReg,
+    /// The label cardinality was inconsistent.
+    #[error("Inconsistent label cardinality, expect {0} label values, but got {1}")]
+    InconsistentCardinality(usize, usize),
```
Comment (breeswish) [inline src/errors.rs]: Maybe we should name it

File: src/errors.rs [hunk 2]
```diff
@@ -11,76 +11,28 @@
...
+    /// An error message which is only a string.
+    #[error("Error: {0}")]
+    Msg(String),
```
Comment (breeswish) [inline src/errors.rs]: In future PRs we should clean up this kind of errors.

File: src/errors.rs [hunk 3]
```diff
@@ -11,76 +11,28 @@
...
+    Io(#[from] std::io::Error),
+    /// An error containing a [`protobuf::Error`].
+    #[cfg(feature = "protobuf")]
+    #[error("Protobuf error: {0}")]
+    Protobuf(#[from] protobuf::error::ProtobufError),
 }
 
 /// A specialized Result type for prometheus.
-pub type Result<T> = result::Result<T, Error>;
+pub type Result<T> = std::result::Result<T, Error>;
```
Comment (breeswish) [inline src/errors.rs]: The size could be a bit large even if it is `Result<()>`..

Comment (koushiro) [reply src/errors.rs]: Yes, do you have any good idea? :(

Comment (breeswish) [reply src/errors.rs]: Nop... @kennytm Ideas?

Comment (kennytm) [reply src/errors.rs]: You could `Box` some components. But you need to define what is meant by "Large" 🙃

**Comment by breeswish** [comment]
@overvenus PTAL

**Pull request merged**